### PR TITLE
Introduce Upgrade Hooks

### PR DESCRIPTION
**Pull request by bastjan**
Allows notifying external systems or running arbitrary code on upgrades.

* Adds the `UpgradeJobHook` kind that allows selecting UpgradeJobs to hook into.
* Special attention has been given to the job tracking, to ensure we track jobs before they get deleted by TTL or other options.
* A OCP simulator can be run with `go run ./sim`.

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Update tests.

### CODE REVIEW

**Review by simu — APPROVED** [review]
LGTM, but maybe you should get a 2nd pair of eyes to go over it.

**Comment by bastjan** [comment]
> LGTM, but maybe you should get a 2nd pair of eyes to go over it.

maybe @HappyTetrahedron or @glrf ? 😊

File: controllers/upgradeconfig_controller.go [hunk 1]
```diff
@@ -134,7 +134,7 @@ findNextRun:
 	latestUpdate := clusterversion.LatestAvailableUpdate(cv)
 	if latestUpdate == nil {
 		l.Info("no updates available")
-		return ctrl.Result{}, r.setLastScheduledUpgrade(ctx, &uc, nextRun)
+		return ctrl.Result{RequeueAfter: 15 * time.Second}, r.setLastScheduledUpgrade(ctx, &uc, nextRun)
```
[the remark this answers is not in the document]
Comment (bastjan) [reply controllers/upgradeconfig_controller.go:137]: not really no. the nice way would be watching the clusterversion here too. But I was lazy and not sure if the complication is worth it. Note this is only reconciled every 15s in an upgrade window and basically a no-op bc. of the cache...

[the remark this answers is not in the document]
Comment (bastjan) [reply controllers/upgradeconfig_controller.go]: Upped to a minute, improved design, and added a comment.

File: controllers/upgradejob_controller.go [hunk 2]
```diff
@@ -349,3 +414,349 @@ func (r *UpgradeJobReconciler) tryLockClusterVersion(ctx context.Context, versio
...
+	slices.Grow(jobStatus, len(jobs.Items))
+	for _, job := range jobs.Items {
+		st, msg := hookStatusFromJob(job)
+		jobStatus = append(jobStatus, managedupgradev1beta1.HookJobTracker{
+			HookEvent:          string(event),
+			UpgradeJobHookName: hook.Name,
+			Status:             st,
+			Message:            msg,
+		})
+	}
```
Comment (glrf) [inline controllers/upgradejob_controller.go:513]: I'm a bit confused by this. When starting reconciliation for the `UpgradeJob` we find all related `Jobs` and write their status in the `HookJobTracker`.

Here we're first fetching the tracked Job info from the `UgradeJob` status and then fetching the status from the actual `Job` and appending it. Aren't we duplicating the tracked `Job` info? Why is this necessary? Am I missing something?

File: api/v1beta1/upgradejobhook_types.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	Spec   UpgradeJobHookSpec   `json:"spec,omitempty"`
+	Status UpgradeJobHookStatus `json:"status,omitempty"`
+}
+
+// Claim claims the hook for the given claimer.
+// Returns true if the hook was claimed, false if it was already claimed.
+// Second return value is true if the hooks status was updated.
+func (u *UpgradeJobHook) Claim(claimer client.Object) (ok, updated bool) {
+	if u.Spec.GetRun() != RunNext {
+		return true, false
```
Comment (glrf) [inline api/v1beta1/upgradejobhook_types.go:132]: Nitpick: I found this a bit confusing at first and didn't get that this was only really relevant for `RunNext` (I mean it's obvious once you actually read the code but still). Maybe adding that to the docstring could help. Or some different naming?

File: controllers/upgradejob_controller.go [hunk 4]
```diff
@@ -349,3 +414,349 @@ func (r *UpgradeJobReconciler) tryLockClusterVersion(ctx context.Context, versio
...
+		if ok, upd := hook.Claim(uj); ok {
+			if upd {
+				l.Info("claimed hook", "hook", hook.Name)
+				if err := r.Status().Update(ctx, &hook); err != nil {
+					return false, err
+				}
+			}
+			hooks = append(hooks, hook)
+		}
+	}
```
Comment (glrf) [inline controllers/upgradejob_controller.go:448]: Nitpick: Splitting this out into a sub-function could help readability a bit.

File: controllers/upgradejob_controller.go [hunk 5]
```diff
@@ -349,3 +414,349 @@ func (r *UpgradeJobReconciler) tryLockClusterVersion(ctx context.Context, versio
...
+
+	activeJobs := []string{}
+	errors := []error{}
+	failedJobs := []string{}
+	for _, hook := range hooks {
+		jobs, err := r.jobForUpgradeJobAndHook(ctx, uj, hook, event)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
```
Comment (glrf) [inline controllers/upgradejob_controller.go:458]: Nitpick: By just reading this I did not get that the `jobForUpgadeJobHook` actually creates the `Job` if it does not exit. I'd probably split fetching job info and actually creating the job, or change the naming of the function

**Review by glrf — COMMENTED** [review]
LGTM overall. Just some nitpicks and I think you're duplicating some job tracker.
